src/main/scala/patmat/Huffman.scala

```diff
@@ -0,0 +1,342 @@
+        Fork(Fork(Leaf('m', 45521), Leaf('p', 46335), List('m', 'p'), 91856),
+             Leaf('u', 96785),
+             List('m', 'p', 'u'),
+             188641),
+        List('o', 'l', 'm', 'p', 'u'),
+        355071
+      ),
+      List('s', 'd', 'x', 'j', 'f', 'z', 'k', 'w', 'y', 'h', 'q', 'o', 'l', 'm', 'p', 'u'),
+      605362
+    ),
+    Fork(
+      Fork(
+        Fork(
+          Leaf('r', 100500),
+          Fork(Leaf('c', 50003),
+               Fork(Leaf('v', 24975),
+                    Fork(Leaf('g', 13288), Leaf('b', 13822), List('g', 'b'), 27110),
+                    List('v', 'g', 'b'),
+                    52085),
+               List('c', 'v', 'g', 'b'),
+               102088),
+          List('r', 'c', 'v', 'g', 'b'),
+          202588
+        ),
+        Fork(Leaf('n', 108812), Leaf('t', 111103), List('n', 't'), 219915),
+        List('r', 'c', 'v', 'g', 'b', 'n', 't'),
+        422503
+      ),
+      Fork(Leaf('e', 225947),
+           Fork(Leaf('i', 115465), Leaf('a', 117110), List('i', 'a'), 232575),
+           List('e', 'i', 'a'),
+           458522),
+      List('r', 'c', 'v', 'g', 'b', 'n', 't', 'e', 'i', 'a'),
+      881025
+    ),
+    List('s', 'd', 'x', 'j', 'f', 'z', 'k', 'w', 'y', 'h', 'q', 'o', 'l', 'm', 'p', 'u', 'r', 'c', 'v', 'g', 'b', 'n',
+      't', 'e', 'i', 'a'),
+    1486387
+  )
+
+  /**
+    * What does the secret message say? Can you decode it?
+    * For the decoding use the 'frenchCode' Huffman tree defined above.
+    */
+  val secret: List[Bit] = List(0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1,
+    1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1)
+
+  /**
+    * Write a function that returns the decoded secret
+    */
+  def decodedSecret: List[Char] = decode(frenchCode, secret)
+
+  // Part 4a: Encoding using Huffman tree
+
+  /**
+    * This function encodes `text` using the code tree `tree`
+    * into a sequence of bits.
+    */
+  def encode(tree: CodeTree)(text: List[Char]): List[Bit] = {
+
+    def findChar(tree: CodeTree, char: Char, acc: List[Bit]): List[Bit] = tree match {
+      case Leaf(x, _) =>
+        if (x == char) acc
+        else Nil
+      case Fork(left, right, _, _) =>
+        val leftList = findChar(left, char, acc :+ 0)
+        if (leftList != Nil) leftList
+        else findChar(right, char, acc :+ 1)
+    }
+
+    if (text.isEmpty) Nil
+    else findChar(tree, text.head, List()) ::: encode(tree)(text.tail)
+  }
+
+  // Part 4b: Encoding using code table
+
+  type CodeTable = List[(Char, List[Bit])]
+
+  /**
+    * This function returns the bit sequence that represents the character `char` in
+    * the code table `table`.
+    */
+  @tailrec def codeBits(table: CodeTable)(char: Char): List[Bit] = table match {
+    case Nil          => throw new NoSuchElementException
+    case head :: tail => if (head._1 == char) head._2 else codeBits(tail)(char)
```

Comment:
You can represent `head` as `(a, b)`. Or whatever makes sense to you. 

And you can use `a` and `b` instead of `head._1` and `head._2`.